### ISSUE TITLE
rmse - rms vs energy

### ISSUE THREAD

**Issue by nullmightybofo** [librosa/librosa#762]
This is not really an issue, but I wonder why the rmse feaure:

https://librosa.github.io/librosa/generated/librosa.feature.rmse.html

is named "rms energy". Afaik rms is a 'linear' value (like an amplitude, or any 'field quantity') while energy, like power, is related to suqared amplitudes.

This discussion here 

https://groups.google.com/forum/#!topic/librosa/TIke2HVJHdc

also states that

"energy is the square-root of power." 

I am not quite sure that is correct. Could someone comment on that?

Here are some links that my be useful:

https://en.wikipedia.org/wiki/Root_mean_square

https://en.wikipedia.org/wiki/Energy_(signal_processing)

https://en.wikipedia.org/wiki/Spectral_density

Thanks and BR

**Comment by lostanlen** [comment]
I am under the impression that "RMSE" means "(square) Root of the Mean Squared Error", not "Mean Squared Energy". The "MSE" terminology is inherited from statistics and machine learning. In this context, the "error" is the signal itself, because it is a difference between the signal and the null signal (silence).
This commit by @bmcfee from Feb 2015, which renamed RMS to RMSE, is relevant to the discussion: 
https://github.com/librosa/librosa/commit/cb6bfe55fa07223f633c1fa5156a26939690b8a1#diff-0424f9e447bc0e640b01e2b7de38875e

**Comment by lostanlen** [comment]
> It's just that I think it should not be named "energy" when it is in fact a "linear" value, as in directly proportional to amplitude, not to squared amplitude.
I think we agree here.

It therefore seems like a doc problem (removing the word "energy" from the docstring) rather than a function naming problem.

**Comment by bmcfee** [comment]
Ok, I'm convinced. :)

I can see it as being an issue of both docs and naming.  It's caused enough confusion over the years that clearly something is strange about it, so a PR to clarify the calculation would be most appreciated.  We could also move or alias rmse back to rms.

**Comment by lostanlen** [comment]
Essentia uses the RMS acronym throughout the code base

**Comment by nullmightybofo** [comment]
Sorry for not responding earlier, I can look into writing a PR.

Basically, it amounts to removing the word "energy" from the docstring, probably renaming the function to rms()? And possibly adding an alias rmse() that directly passes on all arguments to a call of the new rms() function for backwards compatibility?

One more thing that I don't understand. The docstring also says this:
 "However, using a spectrogram will give a more accurate representation of energy over time because its frames can be windowed, ..."

Why the representation is supposed to be more accurate due to the windowing I  don't understand. If I am not completely mistaken, when using identical parameters in both cases (identical windows for slicing the signal, frame and hop sizes), by Parseval's theorem, results should be identical?

However, using rmse() directly in the time domain, this function essentially uses a rectangular window for slicing the signal, and this can't be changed. Libros's stft() function (as an example) uses a hann window by default. So I guess results would be different. Will have to test. I don't know whether people are aware of that, or even care when using this? Maybe it is useless nitpicking ;-)

**Comment by bmcfee** [comment]
> Sorry for not responding earlier, I can look into writing a PR.

Great, thanks!

> Basically, it amounts to removing the word "energy" from the docstring, probably renaming the function to rms()? And possibly adding an alias rmse() that directly passes on all arguments to a call of the new rms() function for backwards compatibility?

Yeah.  We have shim helpers for building aliases and deprecations, see https://github.com/librosa/librosa/blob/master/librosa/util/decorators.py#L14 -- so it should be as simple as saying `rmse = util.moved('librosa.feature.rmse', '0.6.3', '0.7.0')(rms)` after the function has been renamed.  We'll then drop the alias in the next major revision.



> One more thing that I don't understand. The docstring also says this:
> "However, using a spectrogram will give a more accurate representation of energy over time because its frames can be windowed, ..."
> 
> Why the representation is supposed to be more accurate due to the windowing I don't understand. If I am not completely mistaken, when using identical parameters in both cases (identical windows for slicing the signal, frame and hop sizes), by Parseval's theorem, results should be identical, regardless of calculating in the time- or in the frequency-domain?
> 
> However, using rmse() directly in the time domain here, this function essentially uses a rectangular window for slicing the signal, and this can't be changed. Libros's stft() function (as an example) uses a hann window by default. So I guess results would be different. Will have to test. I don't know whether people are aware of that, or even care when using this? Maybe it is useless nitpicking ;-)

You are correct: the time-domain rms calculation is implicitly using a rect window, and Parseval doesn't hold here because we might have a different window function in the stft.

I think the docstring there is due to @carlthome , but I suspect it's meant to be "accuracy" in the sense of not being sensitive to edge effects in the framing, so that the rms is weighted toward the center of the frame (due to windowing).

If that's unclear, I'd welcome an amendment to that docstring to more carefully spell out what's going on.

**Comment by carlthome** [comment]
:+1: on renaming `rmse` to `rms`.

Sorry for the crappy docstring @nullmightybofo. I'm learning as I go. 🙃 

> I think the docstring there is due to @carlthome , but I suspect it's meant to be "accuracy" in the sense of not being sensitive to edge effects in the framing, so that the rms is weighted toward the center of the frame (due to windowing).

Yes, precisely. My bad! But come to think of it, why didn't I just apply a `scipy.signal.window` in the time domain version? Could be a good improvement.

**Comment by carlthome** [comment]
I guess one nice thing about using an _implicit_ rectangular window is computational performance.

**Comment by nullmightybofo** [comment]
Thank you for the clarifications!
It's not a crappy docstring at all. It was in conflict with what I had learned earlier, so I wanted that cleared (for me).

Also I just discoverd this further down the docstring, it clearly states the difference between td and fd:
"Use a STFT window of constant ones and no frame centering to get consistent results with the RMS energy computed from the audio samples y"
Time for humble pie for me :-)

"[...]  I suspect it's meant to be "accuracy" in the sense of not being sensitive to edge effects in the framing, so that the rms is weighted toward the center of the frame (due to windowing)."

This makes it clear to me, thanks.

The use of np.abs() in the calculation is for allowing complex-valued signals, correct?

return np.sqrt(np.mean(np.abs(x)**2, axis=0, keepdims=True))

https://librosa.github.io/librosa/_modules/librosa/feature/spectral.html#rmse

For adding windowing in the time-domain, if necessary / desired, the util.frame() function that is used internally by rmse() doesn't support using a window, so that'd have to be multiplied in afterwards somehow.

**Comment by carlthome** [comment]
> Time for humble pie for me :-)

😘 

> The use of np.abs() in the calculation is for allowing complex-valued signals, correct?

No, audio is real valued but amplitudes go from [-1.0, 1.0] so we take the magnitude envelope via `abs`.

> For adding windowing in the time-domain, if necessary / desired, the util.frame() function that is used internally by rmse() doesn't support using a window, so that'd have to be multiplied onto x, after the splitting.

Yeah, something like (I guess):
```py
window = scipy.signal.windows.hann(frame_length)
window = window[None, :]
frames *= window
```

**Comment by bmcfee** [comment]
I don't think we need to implement windowing in the rms calculation here, at least not in this PR.  If folks want that, they probably have a spectrogram kicking around already, right?  For me, the main reason to use time-domain rms calculation is to avoid the extra memory copy required for windowing.

**Comment by lostanlen** [comment]
@nullmightybofo what is the state of this issue?

**Comment by bmcfee** [comment]
If nobody else wants to do this one, I can hack it over the next few days.  We're getting close to clearing the 0.6.3 milestone, and it would be nice to get that out in 2018.

**Comment by nullmightybofo** [comment]
I will do it this weekend. I guess the larger part will be to get the deprecation warnings etc right, as mentioned above. Sorry for not replying earlier.

**Comment by bmcfee** [comment]
Great, thanks!  If you want to see examples of how the move helper was used previously, you can look through older versions of the code, e.g. [this line](https://github.com/librosa/librosa/blob/5f8f601e45ce69f7192cdb524f10160927397a62/librosa/core/spectrum.py#L747)

**Comment by nullmightybofo** [comment]
Here it is: #801 
Apologies for taking so long.
